src/vs/base/common/async.ts

```diff
@@ -2169,4 +2169,23 @@ export class AsyncIterableSource<T> {
 	}
 }
 
+export function cancellableIterable<T>(iterableOrIterator: AsyncIterator<T> | AsyncIterable<T>, token: CancellationToken): AsyncIterableIterator<T> {
+	const iterator = Symbol.asyncIterator in iterableOrIterator ? iterableOrIterator[Symbol.asyncIterator]() : iterableOrIterator;
+
+	return {
+		async next(): Promise<IteratorResult<T>> {
+			if (token.isCancellationRequested) {
+				return { done: true, value: undefined };
+			}
+			const result = await raceCancellation(iterator.next(), token);
+			return result || { done: true, value: undefined };
```

Comment:
We should probably do a more correct undefined check here so that `false`, `''` etc won't mean cancellation. The race-function does take a default result which you can uniquely check against

Reply:
`result` is an `IteratorResult<T>` which'll always be an object, and only undefined if the `token` was cancelled